designs/variables-for-userdata.md

```diff
@@ -0,0 +1,206 @@
+
+## Not in Scope
+
+* This document is not proposing to be able to toggle on/off the static drift capability for user-defined fields in the `EC2NodeClass` spec. This would not solve the use case of rotating credentials in the `userData` field, since the field is an unstructured string and can have virtually any format.
+
+## Solutions
+
+### Option 1: Introduce Non-Drifting Template Variables
+
+This option introduces a new map field, `userDataVariables` to the `EC2NodeClass` spec for defining key-value pairs that can be substituted into the `userData` field. This field and its variables would be explicitly ignored by the hashing algorithms used in the drift detection mechanism. Variables are referenced in `userData` by bare token names (e.g., AUTH_TOKEN) using simple search-and-replace during launch template creation.
+
+```yaml
+apiVersion: karpenter.k8s.aws/v1beta1
+kind: EC2NodeClass
+metadata:
+  name: default
+spec:
+  amiFamily: Custom
+  # New field for non-drifting variables
+  userDataVariables:
+    AUTH_TOKEN: "rotated-secret-token-value"
+  userData: | # an example of a minimal ignition config userData
+  '{
+    "ignition": {
+      "config": {
+        "merge": [
+          {
+            "httpHeaders": [
+              {
+                "name": "Authorization",
+                "value": "Bearer AUTH_TOKEN"
+              }
+            ],
+            "source": "https://bootstrap.internal/config"
+          }
+
+        ]
+      }
+    }
+  }'
+```
+
+#### Implementation Details
+
+The code's AMIFamily resolver would need to:
+
+1. Parse the `userData` template and identify variable tokens by exact name (e.g., AUTH_TOKEN).
+2. Substitute variables with their current values from `userDataVariables` before computing the launch template at runtime.
+
+The code's drift calculator will need to:
+
+1. When calculating the `EC2NodeClass` `Hash()` function, which is used in drift detection, ignore the `userDataVariables` field by tagging the field with `hash:"ignore"`.
+
+#### Advantages
+
+* Simple to use
+* Flexible to work with any `userData` format (shell scripts, cloud-config, etc.) and any `AMIFamily`.
+* Compliant with current drift static field ignoring mechanisms.
+
+#### Disadvantages
+
+* Only supports simple string substitution.
+* Potentially sensitive data stored in the cluster would still be stored in plaintext in the `EC2NodeClass` spec.
+
+### Option 2: Reference Kubernetes Secrets
+
+This option introduces a new field, `userDataSecretRefs`, to the `EC2NodeClass` spec. This field is a list of references to Kubernetes Secrets, which will be used to populate the `userData` template variables. This is a more secure way to inject dynamic data into the `userData` field, as the secrets are not stored in the `EC2NodeClass` spec in plaintext. This also solves the drift issue, as the secret's underlying data can change, but that is not visible to the drift detection mechanism.
```

Comment:
I think you'd need both options, similar to a pod spec's env var secret refs.  I still have a concern exposing secret refs like this may give a false impression that the path is actually secure, where user-data itself is not encrypted.